c/serverStatusService.c

```diff
@@ -0,0 +1,272 @@
+    return;
+  }
+  jsonPrinter *out = respondWithJsonPrinter(response);
+  setResponseStatus(response, 200, "OK");
+  setDefaultJSONRESTHeaders(response);
+  writeHeader(response);
+  jsonStart(out);
+  jsonAddString(out, "timestamp", c_time_str);
+  if(getenv("ZSS_LOG_FILE") != NULL){
+    jsonAddString(out, "logDirectory", getenv("ZSS_LOG_FILE"));
+  } else {
+    jsonAddString(out, "logDirectory", "ZSS_LOG_FILE not defined in environment variables");
+  }
+  jsonAddString(out, "agentName", "zss");
+  jsonAddString(out, "agentVersion", context->productVersion);
+  jsonAddString(out, "arch", unameRet.sysname);
+  jsonAddString(out, "osRelease", unameRet.release);
+  jsonAddString(out, "osVersion", unameRet.version);
+  jsonAddString(out, "hardwareIdentifier", unameRet.machine);
+  jsonAddString(out, "hostname", hostnameBuffer);
+  jsonAddString(out, "nodename", unameRet.nodename);
+  jsonStartObject(out, "userEnvironment");
+  char *env_var = *environ;
+  for (int i = 1; env_var; i++) {
+    char *equalSign = strchr(env_var, '=');
+    if(equalSign == NULL){
+      break;
+    }
+    int nameLen = strlen(env_var) - strlen(equalSign);
+    int nameBufferLen = nameLen + 1;
+    char *name = safeMalloc(nameBufferLen, "env_var name");
+    memcpy(name, env_var, nameLen);
+    name[nameLen] = '\0';
+    jsonAddString(out, name, equalSign+1);
+    safeFree(name, nameBufferLen);
+    env_var = *(environ+i);
+  }
+  jsonEndObject(out);
+  jsonAddInt(out, "demandPagingRate", demandPaging);
+  jsonAddInt(out, "stdCP_CPU_Util", cpuUtil);
+  jsonAddInt(out, "stdCP_MVS_SRM_CPU_Util", mvsSrm);
+  jsonAddInt(out, "ZAAP_CPU_Util", zaapUtil);
+  jsonAddInt(out, "ZIIP_CPU_Util", ziipUtil);
+  jsonAddInt(out, "PID", getpid());
+  jsonAddInt(out, "PPID", getppid());
+  jsonEnd(out);
+  finishResponse(response);
+}
+
+static int serveStatus(HttpService *service, HttpResponse *response) {
+  HttpRequest *request = response->request;
+  ServerAgentContext *context = service->userPointer;
+  //This service is conditional on RBAC being enabled because it is a sensitive URL that only RBAC authorized users should be able to access
+  JsonObject *dataserviceAuth = jsonObjectGetObject(context->serverConfig, "dataserviceAuthentication");
+  int rbacParm = jsonObjectGetBoolean(dataserviceAuth, "rbac");
+  if(!rbacParm){
+     respondWithError(response, HTTP_STATUS_BAD_REQUEST, "Set dataserviceAuthentication.rbac to true in server configuration");
+  } else {
+    if(!strcmp(request->method, methodGET)){
+      char *l1 = stringListPrint(request->parsedFile, 2, 1, "/", 0);
+      if(!strcmp(l1, "")){
+        respondWithServerRoutes(response);
+      }else if (!strcmp(l1, "config")){
+        respondWithServerConfig(response, context->serverConfig);
+      }else if (!strcmp(l1, "log")) {
+        char* logDir = getenv("ZSS_LOG_FILE");
+        if(logDir == NULL || strcmp(logDir, "")){
+          respondWithUnixFile2(NULL, response, logDir, 0, 0, false);
```

Comment:
You can use respondWithUnixFileContents() which chains to call this function eventually.

Reply:
respondWithUnixFileContents() can't be used because it requires a query to unset asB64 in respondWithUnixFile2().

httpserver.c line 3536:
```
    bool asB64 = TRUE;
    char *base64Param = getQueryParam(response->request, "responseType");

    if (NULL != base64Param) {
      if (!strcmp(strupcase(base64Param), "RAW")) {
        asB64 = FALSE;
      }
    }

    respondWithUnixFile2(service, response, absolutePath, jsonMode, autocvt, asB64);
```

Reply:
OK